addons/stock/models/stock_move.py

```diff
@@ -402,53 +402,57 @@ def _compute_product_availability(self):
 
+        # Prefetch product info to avoid fetching all product fields
+        self.product_id.read(['type', 'uom_id'], load=False)
+
         not_product_moves = self.filtered(lambda move: move.product_id.type != 'product')
         for move in not_product_moves:
             move.forecast_availability = move.product_qty
 
         product_moves = (self - not_product_moves)
-        warehouse_by_location = {loc: loc.warehouse_id for loc in product_moves.location_id}
 
-        outgoing_unreserved_moves_per_warehouse = defaultdict(lambda: self.env['stock.move'])
+        outgoing_unreserved_moves_per_warehouse = defaultdict(set)
+        now = fields.Datetime.now()
+
+        def key_virtual_available(move):
+            return move.location_id.warehouse_id.id, max(move.date, now)
+
+        # Prefetch efficiently virtual_available for _consuming_picking_types draft move.
+        prefetch_virtual_available = defaultdict(set)
+        virtual_available_dict = {}
+        for move in product_moves:
+            if move.picking_type_id.code in self._consuming_picking_types() and move.state == 'draft':
+                prefetch_virtual_available[key_virtual_available(move)].add(move.product_id.id)
+        for key_context, product_ids in prefetch_virtual_available.items():
+            read_res = self.env['product.product'].browse(product_ids).with_context(warehouse=key_context[0], to_date=key_context[1]).read(['virtual_available'])
+            virtual_available_dict[key_context] = {res['id']: res['virtual_available'] for res in read_res}
+
         for move in product_moves:
             if move.picking_type_id.code in self._consuming_picking_types():
                 if move.state == 'assigned':
                     move.forecast_availability = move.product_uom._compute_quantity(
                         move.reserved_availability, move.product_id.uom_id, rounding_method='HALF-UP')
                 elif move.state == 'draft':
                     # for move _consuming_picking_types and in draft -> the forecast_availability > 0 if in stock
-                    warehouse = move.location_id.warehouse_id
-                    next_date = max(move.date, fields.Datetime.now())
-                    move.forecast_availability = move.product_id.with_context(warehouse=warehouse.id, to_date=next_date).virtual_available - move.product_qty
+                    move.forecast_availability = virtual_available_dict[key_virtual_available(move)][move.product_id.id] - move.product_qty
                 elif move.state in ('waiting', 'confirmed', 'partially_available'):
-                    outgoing_unreserved_moves_per_warehouse[warehouse_by_location[move.location_id]] |= move
+                    outgoing_unreserved_moves_per_warehouse[move.location_id.warehouse_id].add(move.id)
             elif move.picking_type_id.code == 'incoming':
                 move.forecast_availability = move._get_forecast_availability_incoming()
```

Comment:
I think you could do the same dict as you did for the draft move to replace _get_forecast_availability_incoming()